### PR TITLE
Support api key auth so we may initiate reindexes against packages

### PR DESCRIPTION
**Pull request by LATaylor-guardian**

### CODE REVIEW

File: public/components/contentSourceEdit.react.js [hunk 1]
```diff
@@ -37,11 +61,17 @@ export default class ContentSourceEdit extends React.Component {
         var appName = this.state.appName.trim();
         var description = this.state.description.trim();
         var reindexEndpoint = this.state.reindexEndpoint.trim();
-        if(!appName || !description || !reindexEndpoint) {
+        var authType = this.state.authType.trim();
+        var apiKey = (this.state.apiKey || '').trim();
+
+        if ("api-key" != authType) apiKey = '';
+        
+        if(!appName || !description || !reindexEndpoint || !authType) {
```
Comment (mchv) [inline public/components/contentSourceEdit.react.js]: could we convert this to:
```js
if (appName && description && reindexEndpoint && authType) {
   ...
} else {
  this.setState({alertStyle: 'danger', alertMessage: 'Invalid form. Correct the fields and try again.', alertVisibility: true});
  return;
}
```

I think it is easier to read, no?

Comment (LATaylor-guardian) [inline public/components/contentSourceEdit.react.js]: Yeah, I'll change it.

File: app/com/gu/floodgate/contentsource/ContentSource.scala [hunk 2]
```diff
@@ -1,13 +1,20 @@
...
  */
 @JsonFormat
-case class ContentSourceWithoutId(appName: String, description: String, reindexEndpoint: String, environment: String)
+case class ContentSourceWithoutId(
+  appName: String,
+  description: String,
+  reindexEndpoint: String,
+  environment: String,
+  authType: String,
+  apiKey: Option[String])
```
Comment (cb372) [inline app/com/gu/floodgate/contentsource/ContentSource.scala]: Does it need to be a separate field? Couldn't the user just include the api key param in the endpoint URL? e.g. `http://my-service/reindex?key=foo`. This means they can name the URL param however they like.

Comment (LATaylor-guardian) [inline app/com/gu/floodgate/contentsource/ContentSource.scala]: Yes you could, I didn't think of this!

File: app/com/gu/floodgate/contentsource/ContentSource.scala [hunk 3]
```diff
@@ -16,13 +23,37 @@ case class ContentSourcesResponse(contentSources: Seq[ContentSource])
...
+    val apiKeyWithQuestionMarkPrefix = apiKeyParamwithPrefix("?").getOrElse("")
+    val apiKeyWithAmpersandPrefix = apiKeyParamwithPrefix("&").getOrElse("")
+
+    (dateParameters.from, dateParameters.to) match {
+      case (None, None) => s"${this.reindexEndpoint}?$apiKeyWithQuestionMarkPrefix"
+      case (Some(f), None) => s"${this.reindexEndpoint}?from=${f.toString}$apiKeyWithAmpersandPrefix"
+      case (None, Some(t)) => s"${this.reindexEndpoint}?to=${t.toString}$apiKeyWithAmpersandPrefix"
+      case (Some(f), Some(t)) => s"${this.reindexEndpoint}?from=${f.toString}&to=${t.toString}$apiKeyWithAmpersandPrefix"
+    }
+  }
```
Comment (cb372) [inline app/com/gu/floodgate/contentsource/ContentSource.scala]: Why are you constructing the URL manually? You can pass the URL params separately from the main URL when you construct the request using Play WS.

If you really want to build it yourself, how about this?

```scala
val apiKey: Option[String] = ...
val from: Option[String] = ...
val to: Option[String] = ...

def param(key: String, value: Option[String]): Option[String] = value.map(v => s"$key=$v")
val queryString: String = Seq(param("api-key", apiKey), param("from", from), param("to", to)).flatten.mkString("&")
s"$reindexEndpoint?$queryString"
```

File: app/com/gu/floodgate/contentsource/ContentSource.scala [hunk 4]
```diff
@@ -16,13 +22,28 @@ case class ContentSourcesResponse(contentSources: Seq[ContentSource])
...
   def uniqueId: String = s"$id-$environment"
+
+  def reindexEndpointWithDateParams(dateParameters: DateParameters): String = {
+
+    def param(key: String, value: Option[String]): Option[String] = value.map(v => s"$key=$v")
+
+    val paramFrom = param("from", dateParameters.from map (_.toString))
+    val paramTo = param("to", dateParameters.to map (_.toString))
+    val queryString: String = Seq(paramFrom, paramTo).flatten.mkString("&")
+    val urlSeparator = if ("api-key" == this.authType) "&" else "?"
```
Comment (cb372) [inline app/com/gu/floodgate/contentsource/ContentSource.scala]: I would change this predicate to `if (reindexEndpoint contains "?")`. The URL might already contain parameters for some reason unrelated to the api key.

**Comment by cb372** [comment]
:+1:

**Pull request merged**